x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java

```diff
@@ -156,18 +160,19 @@ public Void run() {
     // private final Environment env;
     protected final Licensing licensing;
     // These should not be directly accessed as they cannot be overridden in tests. Please use the getters so they can be overridden.
-    private static final SetOnce<XPackLicenseState> licenseState = new SetOnce<>();
     private static final SetOnce<SSLService> sslService = new SetOnce<>();
-    private static final SetOnce<LicenseService> licenseService = new SetOnce<>();
     private static final SetOnce<LongSupplier> epochMillisSupplier = new SetOnce<>();
+    // non-final to allow for testing
+    private static SetOnce<XPackLicenseState> licenseState = new SetOnce<>();
+    private static SetOnce<LicenseService> licenseService = new SetOnce<>();
```

Comment:
This feels off. Do we really need to drop `final`?

Reply:
> This feels off. Do we really need to drop final?

Yeah it is needed for testing. Since these are static and SetOnce, you can only set this once per JVM.  We didn't have any testing before now. Now that we have testing we need to set this more than once per JVM, so I move creation of these objects to the constructor. The way a test can create many instances and this can be set differently per test. A SetOnce only created in the Constructor is pretty close (but not exact) to a final variable. In general promoting class state is static state IMO is what feels off.